Use fast set location at startup

iSazonov opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

At startup we create initial runspace and set default location to current file system location (path).
For that we use _generic_ SetLocation() which designed to work with all providers and do many extra checks/operations.
In the startup scenario we know that the path is _valid_ file system path so we can use more simple and more fast set location.

On my system I get win ~10ms or ~1.62% and ~8Mb or ~3%.
The win comes from removing some extra file operations:
![image](https://user-images.githubusercontent.com/22290914/63431597-a19acc00-c438-11e9-9367-31d51a6e354c.png)

## PR Context

Related #6443

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

iSazonov committed:
Use fast set location at startup

iSazonov committed:
Fix typo

bergmeister commented:
> In the startup scenario we know that the path is valid file system path so we can use more simple and more fast set location.
What about the Azure Cloud Shell, which starts by default in the `Azure` drive?

```
Requesting a Cloud Shell.Succeeded.
Connecting terminal...
Welcome to Azure Cloud Shell

Type "az" to use Azure CLI
Type "help" to learn about Cloud Shell

get-locationMOTD: Discover installed Azure modules: Get-Module Az* -ListAvailableVERBOSE: Authenticating to Azure ...VERBOSE: Building your Azure drive ...
Azure:/
PS Azure:\> Get-Location

Path
----
Azure:/

PS Azure:\> Get-PSDrive

Name           Used (GB)     Free (GB) Provider      Root                                                                                                                                                                    CurrentLocation
----           ---------     --------- --------      ----                                                                                                                                                                    ---------------
/                  19.71         28.70 FileSystem    /                                                                                                                                                                        home/christoph
Alias                                  Alias
Azure                                  SHiPS         AzurePSDrive#Azure
Env                                    Environment
Function                               Function
Variable                               Variable
```

iSazonov commented:
> What about the Azure Cloud Shell, which starts by default in the Azure drive?

I guess it is loaded after initial startup and also it is remoting scenario.

SteveL-MSFT commented:
@bergmeister I'm pretty sure CloudShell uses the PowerShell profile to create the Azure drive, logon, and Set-Location so all that is after pwsh startup anyways.

iSazonov commented:
Perhaps this code can still be improved, but it also works for Linux where I can’t debug :-(

daxian-dbw commented:
@TravisEz13 and @adityapatwardhan Given you two have experiences with FileSystemProvider related code, could you please review this PR when you have time? Thanks!

iSazonov commented on src/System.Management.Automation/engine/SessionStateLocationAPIs.cs:
```diff
@@ -590,6 +590,79 @@ internal PathInfo SetLocation(string path, CmdletProviderContext context, bool l
+            {
+                // See if the path is a relative or absolute
+                // path.
+                if (Globber.IsAbsolutePath(path, out string driveName))
```
I feel this could be optimized too but I don't know how this works on Unix and I afraid to make breaking change.

Additional thought is that the branch works only on Unix and _at startup_ we possibly already initialized current drive and no need to repeat the step.

iSazonov commented on src/System.Management.Automation/engine/SessionStateLocationAPIs.cs:
```diff
@@ -590,6 +590,79 @@ internal PathInfo SetLocation(string path, CmdletProviderContext context, bool l
+            PathInfo current = null;
+            if (PublicSessionState.InvokeCommand.LocationChangedAction != null)
+            {
+                // It is under a condition to avoid the unneeded allocation - CurrentLocation always allocates.
```
I want protect the condition from removing because CurrentLocation always allocates and we should avoid the allocation if we don't use CurrentLocation in code below. Your proposal doesn't reflect this.
What about "CurrentLocation always allocates and we should avoid the allocation if we don't use CurrentLocation later."?

iSazonov committed:
Update src/System.Management.Automation/engine/SessionStateLocationAPIs.cs

Co-Authored-By: Travis Plunk <travis.plunk@microsoft.com>

iSazonov committed:
Remove Globber.IsAbsolutePath() check

iSazonov commented on src/System.Management.Automation/engine/SessionStateLocationAPIs.cs:
```diff
@@ -590,6 +590,79 @@ internal PathInfo SetLocation(string path, CmdletProviderContext context, bool l
+            {
+                // See if the path is a relative or absolute
+                // path.
+                if (Globber.IsAbsolutePath(path, out string driveName))
```
We call the method twice with:
- Directory.GetCurrentDirectory()
- System.IO.Path.GetDirectoryName(PsUtils.GetMainModule(currentProcess).FileName)
The both is absolute file system paths on Windows so that we have no need to reassign current drive - it is the same.

I added a commit with the cleanup.

PoshChan commented on src/System.Management.Automation/engine/InitialSessionState.cs:
```diff
@@ -3259,20 +3259,17 @@ internal static void SetSessionStateDrive(ExecutionContext context, bool setLoca
 
```
@iSazonov, your last commit had 84 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

Clean up open Runspaces when exit powershell process.PowerShell process should not freeze at exit
```powershell
No such file or directory
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Api/BasicEngine.Tests.ps1: line 52
52:         $process = Start-Process pwsh -ArgumentList $command -PassThru -RedirectStandardOutput $outputFile

```
Enable-ExperimentalFeature and Disable-ExperimentalFeature tests.Disable-ExperimentalFeature will disable Experimental Feature for scope: AllUsers
```powershell
Program 'pwsh' failed to run: No such file or directoryAt /home/vsts/work/1/s/test/powershell/engine/ExperimentalFeature/EnableDisable-ExperimentalFeature.Tests.ps1:80 char:20
+ ...  $feature = pwsh -noprofile -output xml -command Get-ExperimentalFeat ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/ExperimentalFeature/EnableDisable-ExperimentalFeature.Tests.ps1: line 80
80:         $feature = pwsh -noprofile -output xml -command Get-ExperimentalFeature ExpTest.FeatureOne

```
Basic SysLog tests on Linux.Verifies basic logging with no customizations
```powershell
Program 'pwsh' failed to run: No such file or directoryAt /home/vsts/work/1/s/test/powershell/Host/Logging.Tests.ps1:173 char:9
+         & $powershell -NoProfile -SettingsFile $configFile -Command ' ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/Logging.Tests.ps1: line 173
173:         & $powershell -NoProfile -SettingsFile $configFile -Command '$env:PSModulePath | out-null'

```
Enable-ExperimentalFeature and Disable-ExperimentalFeature tests.Enable-ExperimentalFeature will enable Experimental Feature for scope: AllUsers
```powershell
Program 'pwsh' failed to run: No such file or directoryAt /home/vsts/work/1/s/test/powershell/engine/ExperimentalFeature/EnableDisable-ExperimentalFeature.Tests.ps1:61 char:20
+ ...  $feature = pwsh -noprofile -output xml -command Get-ExperimentalFeat ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/ExperimentalFeature/EnableDisable-ExperimentalFeature.Tests.ps1: line 61
61:         $feature = pwsh -noprofile -output xml -command Get-ExperimentalFeature ExpTest.FeatureOne

```
Basic SysLog tests on Linux.Verifies scriptblock logging with null character
```powershell
Program 'pwsh' failed to run: No such file or directoryAt /home/vsts/work/1/s/test/powershell/Host/Logging.Tests.ps1:229 char:17
+ ...     $null = & $powershell -NoProfile -SettingsFile $configFile -Comma ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/Logging.Tests.ps1: line 229
229:         $null = & $powershell -NoProfile -SettingsFile $configFile -Command $testScriptPath

```

PoshChan replied:
@iSazonov, your last commit had 91 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

Clean up open Runspaces when exit powershell process.PowerShell process should not freeze at exit
```powershell
This command cannot be run due to the error: The directory name is invalid..
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Api\BasicEngine.Tests.ps1: line 52
52:         $process = Start-Process pwsh -ArgumentList $command -PassThru -RedirectStandardOutput $outputFile

```
Cdxml cmdlets are supported.Module level tests.The CimTest module should have the proper cmdlets
```powershell
MofComp.exe failed with exit code 
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Cdxml\Cdxml.Tests.ps1: line 98
98:             throw "MofComp.exe failed with exit code $MofCompReturnCode"

```
Cdxml cmdlets are supported.Get-CimTest cmdlet.The Get-CimTest cmdlet should return 4 objects
```powershell
MofComp.exe failed with exit code 
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Cdxml\Cdxml.Tests.ps1: line 98
98:             throw "MofComp.exe failed with exit code $MofCompReturnCode"

```
Cdxml cmdlets are supported.Get-CimTest cmdlet.The Get-CimTest cmdlet should retrieve an object via id
```powershell
MofComp.exe failed with exit code 
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Cdxml\Cdxml.Tests.ps1: line 98
98:             throw "MofComp.exe failed with exit code $MofCompReturnCode"

```
Cdxml cmdlets are supported.Get-CimTest cmdlet.The Get-CimTest cmdlet should retrieve an object by piped id
```powershell
MofComp.exe failed with exit code 
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Cdxml\Cdxml.Tests.ps1: line 98
98:             throw "MofComp.exe failed with exit code $MofCompReturnCode"

```

iSazonov committed:
Revert "Remove Globber.IsAbsolutePath() check"

This reverts commit c264379f0b701f8f4aaae74255e74b50126eaa95.

daxian-dbw reviewed (commented):
I'm generally concerned about the risk of changing the provider code for perf purpose, especially when the gain is insignificant, as the code can be exercised in so many different ways, it's hard for us to enumerate all the scenarios that might be affected.
I think it's worth for the committee to take this PR as case for discussion, to think about what the guidelines/balance should be for performance changes and sensitive code.

joeyaiello commented:
@PowerShell/powershell-committee reviewed this one, and unfortunately we think we're going to have to reject it. 

* We're concerned about the risk of unknown behavioral changes. We just don't know enough about what this does to make a statement that this is an okay change to make. For example, as an internal API, other parts of the code could be depending on the non-fast path. 
* The performance impact is really too negligible for it to be worth the risk: 1.6% startup time is statistically insignificant, and any RAM savings is going to be gone as soon as PowerShell does anything with the filesystem.

iSazonov commented on src/System.Management.Automation/engine/SessionStateLocationAPIs.cs:
```diff
@@ -590,6 +590,79 @@ internal PathInfo SetLocation(string path, CmdletProviderContext context, bool l
+                }
+            }
+
+            string currentPath = path.Substring(CurrentDrive.Root.Length);
```
If current drive is Temp:/ we always call IsAbsolutePath() and always set current drive based on the `path`. I checked this for shorter code path that is ResetRunspaceState(). Other code paths call previously Bind(), there is a call SetSessionStateDrive(), then again SetSessionStateDrive() and we fall in IsAbsolutePath() too.

daxian-dbw replied:
See the following repro:
```powershell
cd temp:\
mkdir test
$t =$ExecutionContext.SessionState.GetType()
$p = $t.GetProperties("nonpublic, instance")
$p = $p[0]
$v = $p.GetValue($ExecutionContext.SessionState)
$t = $v.GetType()

## Get the existing 'SetLocation(string)' method
$m1 = $t.GetMethod("SetLocation", [System.Reflection.BindingFlags]"nonpublic, instance", $null, [type[]]@([string]), $null)
## Success. Location is set to 'Temp:\test'
$m1.Invoke($v, @("Temp:\test"))

## Get the new 'SetLocationFileSystemFast' method and invoke it.
$m2 = $t.GetMethod("SetLocationFileSystemFast", [System.Reflection.BindingFlags]"nonpublic, instance")
## Fail with 'ArgumentOutOfRangeException'
$m2.Invoke($v, @("Temp:\test"))
> Exception calling "Invoke" with "2" argument(s): "startIndex cannot be larger than length of string. (Parameter 'startIndex')"
```
I know you have examined the current existing call sites of this method and know they are safe, but the fast-code-path method is internal, and it could be used by someone in other places in future, without knowing the prerequisites or assumptions for this method to work. This is not the right pattern to follow, it will results code hard to maintain.

iSazonov commented:
> For example, as an internal API, other parts of the code could be depending on the non-fast path.

It is not true. We always call the code with `Directory.GetCurrentDirectory()` or `System.IO.Path.GetDirectoryName(PsUtils.GetMainModule(currentProcess).FileName)` that is always a file system file in predictable form. _No other entry points exist_. So there is no unpredictable risk - we can review all code paths.

> The performance impact is really too negligible for it to be worth the risk

My intention was and remains to go through all startup code and remove delays where possible. I would not be surprised if in the end we would get a 20-30 percent increase in performance. (Is startup scenario important?)
After the PR we could optimize the code even more.
We call SetSessionStateDrive() 4 times at startup. We could call only 2 times.
In SetSessionStateDrive() we set current drive and then set location that do again set current drive. We could look if we can avoid the duplication.

Is this enough to continue the work?

daxian-dbw commented:
> No other entry points exist. So there is no unpredictable risk - we can review all code paths.

Again, no other entry points exists today doesn't mean it won't in future, because the new API is internal. This is not a good pattern to follow.

> We call SetSessionStateDrive() 4 times at startup. We could call only 2 times.

This was already done by #10416, wasn't it? And that PR has been accepted.

iSazonov commented:
I rebased to get #10416 in the branch only now. :-) I see we set a location only once at startup.
I believe that we will not be able to speed up the startup scenario by doing something in one place. I think we will have to review all startup code and make small improvements to many subsystems. The set location code is a place where we can still get perf win in some ms. This will seed up creation of runspaces in all scenarios too.
I added new commit where made the method private and move to SessionStateInternal class with SetLocationStateDrive() method. I hope this addresses your concern about internal API.

There is still a comment about AppContainer and I don't still understand how address this after removing globbing code. It seems we need to check that Directory.GetCurrentDirectory() exists?

Thank you for great comments and help!

daxian-dbw commented:
> SessionStateInternal class with SetLocationStateDrive() method. I hope this addresses your concern about internal API.

My concerns are not just it being internal, but in general that it's risky to change the provider code. We cannot enumerate all possible scenarios that exercise the code path. If I can quickly find one potential problem just by looking at the code (which I'm not familiar with), then that's a signal that there may be more issues with the changes that are unknown yet.

My personal principles on performance work are:
1. the work should be well-guided by profiling;
2. it's either low hanging fruits that can be easily reviewed with low risks, or more fundamental algorithm level changes that provide significant perf gain; 
3. the most important one -- no behavioral changes should be introduced for a perf changes, unless it's an approved one.

iSazonov commented:
@daxian-dbw Thanks! Sorry if I noisy.

In the PR we do not do behavioral changes in the PR. 
SetSessionStateDrive() method code and comments say that we use _always_ and _exclusively_ FileSystem provider to set init drive and always set location to Directory.GetCurrentDirectory() if requested.
It seems the method name is misleading. Perhaps InitSessionStateDriveToFileSystemDrive() would be better.
Moreover, this code is enclosed in try-catch and all exceptions is ignored. Even if something happens in the original code, we probably don’t even see it too. We see only the consequences that is another current location than Directory.GetCurrentDirectory() but we have also another path for AppContainer which implies that no script should depend on this.

There are only two scenarios where we set a _folder_ location (that only the PR care).
First, runspace creation scenario, we will see wrong prompt in console. Currently it works well.
Then, we will have init file system location not in Directory.GetCurrentDirectory(). I set current drive to Temp:/ and create new runspace - this works well too. We could add the test. 
Second, there is a runspace reset scenario. I can not found how could we test this but at first look it is the same as runspace creation scenario - in the sense that if creation works, then the reset works too.

Last notice is that single scenario where set location is important is startup one because we want get right prompt. Remaining is not important because we use OS current path for initialization and it is not predictable and reliable source in multi runspace environment which means that nobody should lean on the value and set needed location explicitly in current runspace.

Now мy only concern is with AppContainer scenario, which is mentioned in the comments. I do not understand how this can happen and how debug it. It looks like Directory.GetCurrentDirectory() returns nonexistent value in AppContainer.

iSazonov commented:
@joeyaiello @daxian-dbw 
I added some commits to make the code more clean and understandable and measure a runspace creation scenario with follow script:
```powershell
measure-Command {
for( $i=0; $i -lt 100; $i++) {
    [runspace] $rs1 = [runspacefactory]::CreateRunspace()
    [runspace] $rs2 = [runspacefactory]::CreateRunspace()
    [runspace] $rs3 = [runspacefactory]::CreateRunspace()
    [runspace] $rs4 = [runspacefactory]::CreateRunspace()
    [runspace] $rs5 = [runspacefactory]::CreateRunspace()
    [runspace] $rs6 = [runspacefactory]::CreateRunspace()
    [runspace] $rs7 = [runspacefactory]::CreateRunspace()
    [runspace] $rs8 = [runspacefactory]::CreateRunspace()
    [runspace] $rs9 = [runspacefactory]::CreateRunspace()
    [runspace] $rs10 = [runspacefactory]::CreateRunspace()

    $rs1.Open()
    $rs2.Open()
    $rs3.Open()
    $rs4.Open()
    $rs5.Open()
    $rs6.Open()
    $rs7.Open()
    $rs8.Open()
    $rs9.Open()
    $rs10.Open()

    $rs1.Dispose()
    $rs2.Dispose()
    $rs3.Dispose()
    $rs4.Dispose()
    $rs5.Dispose()
    $rs6.Dispose()
    $rs7.Dispose()
    $rs8.Dispose()
    $rs9.Dispose()
    $rs10.Dispose()
}

}
```

Before the change:
```
Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 18
Milliseconds      : 510
Ticks             : 185101449
TotalDays         : 0.000214237788194444
TotalHours        : 0.00514170691666667
TotalMinutes      : 0.308502415
TotalSeconds      : 18.5101449
TotalMilliseconds : 18510.1449
```

After the change:
```
Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 10
Milliseconds      : 654
Ticks             : 106545000
TotalDays         : 0.000123315972222222
TotalHours        : 0.00295958333333333
TotalMinutes      : 0.177575
TotalSeconds      : 10.6545
TotalMilliseconds : 10654.5
```

That is 41.5% for the samples (~8 ms).
Perhaps this is not very much for startup scenario (~2%) but clearly visible for scenarios that actively use runspaces.

daxian-dbw commented:
Interesting. I will take a close look this week. Currently working on driving the preview.4 release.

iSazonov commented:
PerfView screenshot
![image](https://user-images.githubusercontent.com/22290914/65400222-6e9e7c00-ddda-11e9-9502-af70da80267d.png)

Bind takes 26.8% (that is nearly 100% in AutomationEngine constructor) including UpdatesType() - 13.4% and SetSessionDrive() - 12.0.  The both is good candidates for perf investigations.

daxian-dbw commented:
> I added some commits to make the code more clean and understandable ...

@iSazonov I don't see new commits here. Are there new commits in your branch that are not shown here?

iSazonov commented:
@daxian-dbw Closed PR doesn't show new commits. :-)

iSazonov commented:
@daxian-dbw @SteveL-MSFT I'd want to bring back the change. I still believe it is useful for startup and related scenarios.

daxian-dbw commented:
@PoshChan Please remind me in 24 hours

PoshChan commented:
@daxian-dbw, this is the reminder you requested 24 hours ago
